src/Coordination/KeeperSnapshotManager.cpp

```diff
@@ -336,80 +338,139 @@ void KeeperStorageSnapshot::deserialize(SnapshotDeserializationResult & deserial
         return node.getData().empty() && node.stat == Coordination::Stat{};
     };
 
+    using PathNodePair = std::pair<std::string, KeeperStorage::Node>;
+    using PathNodePairPtr = std::shared_ptr<PathNodePair>;
+
+    ConcurrentBoundedQueue<PathNodePairPtr> node_queue(snapshot_container_size);
+    std::atomic<bool> force_stop = false;
+    auto node_load_thread = ThreadFromGlobalPool([&]
```

Comment:
So, the idea is to simply read the snapshot in one thread and modify the storage in another, right?

I was monitoring this PR because AFAIR stress tests sometimes suffers due to keeper do not startup fast enough, this is especially the case when the server had been terminated abnormally (hence no snapshot had been written), but I don't think this will help, though still worth it (at least from my point of view).

I was thinking about something like `create_snapshot_interval` can be added (in addition to `create_snapshot_on_exit`), to do snapshots periodically to make the initial load faster.

@antonio2368 What do you think?

Reply:
> AFAIR stress tests sometimes suffers due to keeper do not startup fast enough

That was true before but I made some changes to avoid those issues and don't remember that we had those problems recently. If you find such examples, please forward them to me.

> I don't think this will help, though still worth it

Yes, I didn't notice major improvements, but it was better so we don't lose much.


About `create_snapshot_interval`, it's an interesting idea but it would be a bit hard to introduce it as bg job because currently NuRaft simply creates snapshots as part of the commit. Maybe on each commit check the distance, if it's not large enough for snapshot, check how old is the last snapshot or (last log not in snapshot?).

Reply:
>I made some changes to avoid those issues

Great! I don't remember that I've seen it lately, thought maybe I was not looking into tests results that match as before.
Any particular patch/PR that fixes such issues for the reference?

Reply:
https://github.com/ClickHouse/ClickHouse/pull/40627
https://github.com/ClickHouse/ClickHouse/pull/42926

This 2 PRs should've fixed the problem with startup - almost all of the changes are simply eliminating my bad quadratic complexity code 😢